src/collections.h

```diff
@@ -119,6 +128,34 @@ Collection<T>::fold(std::function<T(T, T)> func) {
 
 
 template<typename T>
+template<typename U>
+Collection<std::pair<T, U>>
+Collection<T>::zip(Collection<U> other_list) {
+    // TODO: list size checking 
+    using return_type = std::pair<T, U>;
+
+    std::vector<return_type> list(Data.size());
+    for (int i = 0; i < Data.size(); i++)
+        list[i] = std::make_pair(Data[i], other_list[i]);
+    return Collection<return_type>(list);
+};
+
+
+template<typename T>
+template<typename Function, typename U>
+Collection<typename std::result_of<Function(T, U)>::type>
```

Comment:
Shouldn't this be a pair/tuple? This relies on the function passed to return a pair/tuple.

I thought the function is `T -> U` then the result is a pair of `<T, U>`?